Code/Tools/AssetProcessor/native/utilities/BuilderList.cpp

```diff
@@ -0,0 +1,164 @@
+                return;
+            }
+
+            m_createJobsBuilder = AZStd::move(builder);
+        }
+        else
+        {
+            m_builders.emplace(builder->GetUuid(), builder);
+        }
+    }
+
+    AZStd::shared_ptr<Builder> BuilderList::Find(AZ::Uuid uuid)
+    {
+        if (m_createJobsBuilder && m_createJobsBuilder->GetUuid() == uuid)
+        {
+            return m_createJobsBuilder;
+        }
+
+        auto itr = m_builders.find(uuid);
+
+        if (itr != m_builders.end())
+        {
+            return itr->second;
+        }
+
+        return nullptr;
+    }
+
+    BuilderRef BuilderList::GetFirst(BuilderPurpose purpose)
+    {
+        if (purpose == BuilderPurpose::CreateJobs)
+        {
+            if (m_createJobsBuilder)
+            {
+                if (!m_createJobsBuilder->m_busy)
+                {
+                    m_createJobsBuilder->PumpCommunicator();
+
+                    if (m_createJobsBuilder->IsValid())
+                    {
+                        return BuilderRef(m_createJobsBuilder);
+                    }
+
+                    m_createJobsBuilder = nullptr;
+                }
+                else
+                {
+                    AZ_Warning(
+                        "BuilderList",
+                        false,
+                        "CreateJobs builder requested but existing builder is already busy.  There should not be multiple parallel "
+                        "requests for CreateJobs builders");
+                }
+            }
+
+            return {};
+        }
+
+        for (auto itr = m_builders.begin(); itr != m_builders.end();)
+        {
+            auto& builder = itr->second;
+
+            if (!builder->m_busy)
+            {
+                builder->PumpCommunicator();
+
+                if (builder->IsValid())
+                {
+                    return BuilderRef(builder);
+                }
+
+                itr = m_builders.erase(itr);
+            }
+            else
+            {
+                ++itr;
+            }
+        }
+
+        return {};
+    }
+
+    AZStd::string BuilderList::RemoveByConnectionId(AZ::u32 connId)
+    {
+        AZStd::string uuidString;
+
+        if (m_createJobsBuilder && m_createJobsBuilder->GetConnectionId() == connId)
+        {
+            uuidString = m_createJobsBuilder->UuidString();
+            m_createJobsBuilder->m_connectionId = 0;
+            m_createJobsBuilder = nullptr;
+
+            return uuidString;
+        }
+        else
+        {
+            for (auto itr = m_builders.begin(); itr != m_builders.end(); ++itr)
+            {
+                auto& builder = itr->second;
+
+                if (builder->GetConnectionId() == connId)
+                {
+                    uuidString = builder->UuidString();
+                    builder->m_connectionId = 0;
+                    m_builders.erase(itr);
```

Comment:
need to update itr when erasing? Also aren't we done if we've removed the builder?

Reply:
The return is one line below :)